### PR TITLE
[Badger Batch] Insert approvals with badger batch update

### PR DESCRIPTION
**Pull request by zhangchiqing**
Before migrating to pebble based storage, we first refactor the existing badger based storage to use batch updates instead of transaction, so that the database operations are similar to how pebble stores data.

Referred to https://github.com/onflow/flow-go/pull/6374 for ensuring concurrency-safety.

The following concurrency tests passed:
```
go test --failfast --tags=relic  -run=TestApprovalStoreTwoDifferentApprovalsConcurrently -count=100
PASS
ok      github.com/onflow/flow-go/storage/badger        5.818s
```

### CODE REVIEW

File: storage/badger/operation/reader_batch_writer.go
Comment (zhangchiqing) [inline storage/badger/operation/reader_batch_writer.go:128]: This is a big challenging to implement concurrent-safe, will implement in a separate PR.  As this module doesn't need it yet, I'm leaving it as a TODO item.

Comment (jordanschalm) [reply storage/badger/operation/reader_batch_writer.go:128]: I'm wondering if we even need to implement it? I don't see any usage of `DeleteRange` in our code-base right now.

File: storage/badger/approvals.go
Comment (jordanschalm) [inline storage/badger/approvals.go:75]:
```suggestion
// CAUTION: Caller must acquire `indexing` lock.
func (r *ResultApprovals) index(resultID flow.Identifier, chunkIndex uint64, approvalID flow.Identifier) func(storage.BadgerReaderBatchWriter) error {
```

File: storage/badger/operation/reader_batch_writer.go
Comment (zhangchiqing) [reply storage/badger/operation/reader_batch_writer.go:128]: OK, I can remove it for now.

We need [the DeleteRange method for pebble](https://github.com/onflow/flow-go/blob/leo/v0.33-pebble-storage-to-review/storage/pebble/operation/common.go#L333) to [remove all events](https://github.com/onflow/flow-go/blob/leo/v0.33-pebble-storage-to-review/storage/pebble/operation/events.go#L42), transactions etc for a given block when rolling back executed height. 

I used Pebble's DeleteRange method as it can iterate range of keys with certain prefix and delete them atomically.

File: storage/badger/cache_b.go
Comment (jordanschalm) [inline storage/badger/cache_b.go:90]:
```suggestion
	return func(reader storage.Reader) (V, error) {
```
Far from the highest priority, but it would be great if we could replace "transaction" terminology with "reader" or "writer" terminology where we notice it in new code.

File: storage/badger/operation/common.go
Comment (jordanschalm) [inline storage/badger/operation/common.go:47]: Godoc please

File: storage/badger/operation/reader_batch_writer.go
Comment (jordanschalm) [inline storage/badger/operation/reader_batch_writer.go:82]:
```suggestion
// ToReader is a helper function to convert a *badger.DB to a Reader
```
I think? The input is a DB, so this makes more sense to me.

File: storage/badger/operation/reader_batch_writer.go
Comment (zhangchiqing) [reply storage/badger/operation/reader_batch_writer.go:34]: Implemented storage.OnCommitSucceed.

File: storage/badger/operation/reader_batch_writer.go
Comment (zhangchiqing) [reply storage/badger/operation/reader_batch_writer.go:24]: Makes sense. 

I originally combined them because there is a few places need intermingling reads. But I think we probably are able to refactor those logic, such as bootstrapping a sealing segment.

File: storage/badger/operation/approvals.go
Comment (jordanschalm) [inline storage/badger/operation/approvals.go:25]:
```suggestion
// is only one approval per chunk.
// CAUTION: Use of this function must be synchronized by storage.ResultApprovals.
```
I'd like to communicate which low-level functions must be synchronized by a higher-level storage procedure. Adding documentation and using `Unsafe...` naming seems like a relatively easy way to do that.

File: storage/badger/operation/approvals.go
Comment (jordanschalm) [inline storage/badger/operation/approvals.go:26]:
```suggestion
func UnsafeIndexResultApproval(resultID flow.Identifier, chunkIndex uint64, approvalID flow.Identifier) func(storage.Writer) error {
```

File: storage/badger/operation/reader_batch_writer.go
Comment (jordanschalm) [inline storage/badger/operation/reader_batch_writer.go:24]: Before merging, please add documentation for these and other public methods. I wrote a version in this comment for the reader/writer methods: https://github.com/onflow/flow-go/pull/6381#discussion_r1727719819.

**Review by jordanschalm — COMMENTED** [review]
This looks great as a starting point for implementing read-committed concurrency safety against Badger. Nice work. 

Added a few final comments:
- Before merging this, please add documentation to exported methods and types. 
- I added a suggestion for a pattern to communicate which low-level storage operations must be synchronized by a higher-level lock.

**Review by jordanschalm — APPROVED** [review]
Approving to unblock as I'll be away next week and I don't think my outstanding comments require further discussion. I think this is broadly a good starting point for the plan to implement read-committed concurrency safety against Badger. 

I'm happy to merge this once remaining comments have been addressed (mainly documentation), and we make the decision to go ahead with this plan to implement concurrency safety changes against Badger.

**Comment by zhangchiqing** [comment]
Close for now. Found a better way to refactor.

**Comment by zhangchiqing** [comment]
In favor of https://github.com/onflow/flow-go/pull/6466

**Pull request closed without merging**
